File: app/src/main/java/com/eje_c/reastream/MainActivity.java
package com.eje_c.reastream;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.RadioGroup;

import java.net.UnknownHostException;

@SuppressWarnings("ConstantConditions")
public class MainActivity extends AppCompatActivity {
    private ReaStream reaStream;
    private SharedPreferences prefs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        reaStream = new ReaStream();

        prefs = getSharedPreferences("app_status", MODE_PRIVATE);

        EditText identifier = (EditText) findViewById(R.id.input_identifier);
        identifier.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                final String newVal = s.toString();
                prefs.edit().putString("identifier", newVal).apply();
                reaStream.setIdentifier(newVal);
            }
        });

        final String defaultIdentifier = prefs.getString("identifier", null);
        if (defaultIdentifier != null) {
            identifier.setText(defaultIdentifier);
        }

        RadioGroup radioGroup = (RadioGroup) findViewById(R.id.radioGroup_mode);
        radioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                switch (checkedId) {
                    case R.id.radio_mode_receive:
                        if (reaStream.isSending()) {
                            reaStream.stopSending();
                        }
                        if (!reaStream.isReceiving()) {
                            reaStream.startReveiving();
                        }
                        break;
                    case R.id.radio_mode_send:
                        if (reaStream.isReceiving()) {
                            reaStream.stopReceiving();
                        }
                        if (!reaStream.isSending()) {
                            reaStream.startSending();
                        }
                        break;
                }
            }
        });

        CheckBox enabled = (CheckBox) findViewById(R.id.check_enabled);
        enabled.setChecked(reaStream.isEnabled());
        enabled.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                reaStream.setEnabled(isChecked);
            }
        });

        reaStream.startReveiving();

        EditText remoteAddress = (EditText) findViewById(R.id.input_remoteAddress);
        remoteAddress.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                final String newVal = s.toString();

                // Check IP address format
                if (newVal.matches("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}")) {
                    prefs.edit().putString("remoteAddress", newVal).apply();
                    try {
                        reaStream.setRemoteAddress(newVal);
                    } catch (UnknownHostException e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        final String defaultRemoteAddress = prefs.getString("remoteAddress", null);
        if (defaultRemoteAddress != null) {
            remoteAddress.setText(defaultRemoteAddress);
        }

        try {
            reaStream.setRemoteAddress(remoteAddress.getText().toString());
        } catch (UnknownHostException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onDestroy() {
        reaStream.close();
        super.onDestroy();
    }
}

File: app/src/main/java/com/eje_c/reastream/MainActivity.kt
package com.eje_c.reastream

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.TextWatcher
import com.eje_c.libreastream.ReaStream
import kotlinx.android.synthetic.main.activity_main.*
import java.net.UnknownHostException

class MainActivity : AppCompatActivity() {

    private var reaStream: ReaStream = ReaStream()
    private lateinit var prefs: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Init lateinit fields
        prefs = getSharedPreferences("app_status", Context.MODE_PRIVATE)

        // Watch identifier text field
        identifier.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                val newVal = s.toString()

                // Save identifier value to SharedPreference
                prefs.edit().putString("identifier", newVal).apply()

                // Set identifier
                updateReaStream(identifier = newVal)
            }
        })

        // Set previously saved identifier
        val defaultIdentifier = prefs.getString("identifier", null)
        if (defaultIdentifier != null) {
            identifier.setText(defaultIdentifier)
            updateReaStream(identifier = defaultIdentifier)
        }

        // Switch ReaStream mode
        radioGroup.setOnCheckedChangeListener { _, checkedId ->
            when (checkedId) {
                R.id.radio_mode_receive -> {
                    reaStream.stopSending()
                    reaStream.startReceiving()
                }
                R.id.radio_mode_send -> {
                    reaStream.stopReceiving()
                    reaStream.startSending()
                }
            }
        }

        // Bind "enabled" checkbox to ReaStream.isEnabled
        enabled.isChecked = reaStream.isEnabled
        enabled.setOnCheckedChangeListener { _, isChecked -> reaStream.isEnabled = isChecked }

        // Default is receiving mode
        reaStream.startReceiving()

        // Watch remote address text field
        remoteAddress.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                val newVal = s.toString()

                // Check IP address format
                if (newVal.matches("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}".toRegex())) {

                    try {
                        reaStream.setRemoteAddress(newVal)

                        // Save
                        prefs.edit().putString("remoteAddress", newVal).apply()
                    } catch (e: UnknownHostException) {
                        e.printStackTrace()
                    }
                }
            }
        })

        // Set previously saved remote address
        val defaultRemoteAddress = prefs.getString("remoteAddress", null)
        if (defaultRemoteAddress != null) {
            remoteAddress.setText(defaultRemoteAddress)

            try {
                reaStream.setRemoteAddress(defaultRemoteAddress)
            } catch (e: UnknownHostException) {
                e.printStackTrace()
            }
        }

        // Show MIDI events
        reaStream.onMidiEvents = { midiEvents ->

            // This callback is called in background thread.
            // UI update must be in main thread.

            runOnUiThread {
                this.midiEvents.text = midiEvents.map { ev -> ev.toString() }.joinToString()
            }
        }
    }

    private fun updateReaStream(identifier: String = reaStream.identifier) {

        // Original state
        val receiving = reaStream.isReceiving
        val sending = reaStream.isSending
        val remoteAddress = reaStream.remoteAddress

        // Close current session
        reaStream.close()

        // Create new instance
        reaStream = ReaStream(identifier = identifier)

        // Restore previous state
        reaStream.remoteAddress = remoteAddress

        if (receiving) {
            reaStream.startReceiving()
        }

        if (sending) {
            reaStream.startSending()
        }
    }

    override fun onDestroy() {
        reaStream.close()
        super.onDestroy()
    }
}
